File: 01_Abstract_Factory/cpp/src/main.cpp
#include <iostream>
#include <memory>
#include <string>

// Abstracted Product
class Button{
public:
    virtual ~Button() = default;
    virtual void paint () const = 0;
};

class CheckBox{
public:
    virtual ~CheckBox() = default;
    virtual void paint() const = 0;
};

// Abstracted Factory
class GuiFactory{
public:
    virtual ~GuiFactory() = default;
    virtual std::unique_ptr<Button> createButton () const = 0;
    virtual std::unique_ptr<CheckBox> createCheckBox() const = 0;
};

// Concrete Products: Windows
class WindowsButton : public Button {
public:
    void paint() const override {
        std::cout << "Windows Button\n";
    }
};

class WindowsCheckBox : public CheckBox {
public:
    void paint() const override {
        std::cout << "Windows CheckBox\n";
    }
};

// Concrete Product : Mac
class MacButton : public Button {
public:
    void paint() const override {
        std::cout << "Mac Button\n";
    }
};

class MacCheckBox : public CheckBox {
public: 
    void paint() const override {
        std::cout << "Mac CheckBox\n";
    }
};

// Concrete Factories
class WindowsFactory : public GuiFactory {
public:
    std::unique_ptr<Button> createButton() const override {
        return std::make_unique<WindowsButton>();
    }
    std::unique_ptr<CheckBox> createCheckBox() const override {
        return std::make_unique<WindowsCheckBox>();
    }
};

class MacFactory : public GuiFactory {
public: 
    std::unique_ptr<Button> createButton() const override {
        return std::make_unique<MacButton>();
    }
    std::unique_ptr<CheckBox> createCheckBox() const override {
        return std::make_unique<MacCheckBox>();
    }
};

// Client code
void renderUI(const GuiFactory& factory) {
    auto button = factory.createButton();
    auto checkbox = factory.createCheckBox();

    button->paint();
    checkbox->paint();
}

int main() {
    std::string os = "windows";
    std::unique_ptr<GuiFactory> factory;

    if (os == "mac") {
        factory = std::make_unique<MacFactory>();
    } else {
        factory = std::make_unique<WindowsFactory>();
    }

   renderUI(*factory);
}

File: 01_Abstract_Factory/rs/src/main.rs
/*
 * Abstract Factory Pattern (Approach 2: Enum / Parameterized Factory)
 *
 * In idiomatic Rust, instead of dynamic dispatch with traits and heap allocations (`Box<dyn Trait>`),
 * we can leverage Rust's powerful enum system:
 * - Avoids heap allocations (`Box`) and vtable indirection.
 * - Entirely stack-allocated and resolved via static dispatch.
 * - The compiler enforces exhaustive pattern matching across all OS variants.
 */

// 1. Family / Variant Enum
// Represents the distinct families of related products (e.g., Windows vs Mac).
#[derive(Clone, Copy, Debug, PartialEq)]
enum Os {
    Windows,
    Mac,
}

// 2. Concrete Products
// Products are parameterized by the OS variant and match on it during operations.
struct Button {
    os: Os,
}

impl Button {
    fn paint(&self) {
        match self.os {
            Os::Windows => println!("Rendering Windows button"),
            Os::Mac => println!("Rendering Mac button"),
        }
    }
}

struct Checkbox {
    os: Os,
}

impl Checkbox {
    fn paint(&self) {
        match self.os {
            Os::Windows => println!("Rendering Windows checkbox"),
            Os::Mac => println!("Rendering Mac checkbox"),
        }
    }
}

// 3. Concrete Factory
// The factory stores the selected OS family and produces consistent products for that family.
struct GuiFactory {
    os: Os,
}

impl GuiFactory {
    fn new(os: Os) -> Self {
        Self { os }
    }

    fn create_button(&self) -> Button {
        Button { os: self.os }
    }

    fn create_checkbox(&self) -> Checkbox {
        Checkbox { os: self.os }
    }
}

// 4. Client Code
// Operates on the factory to produce and render a cohesive family of UI components.
fn render_factory(factory: &GuiFactory) {
    let button = factory.create_button();
    let checkbox = factory.create_checkbox();

    button.paint();
    checkbox.paint();
}

fn main() {
    let os_str = "windows";

    // Select the product family based on runtime configuration
    let os = match os_str {
        "windows" => Os::Windows,
        "mac" => Os::Mac,
        _ => panic!("Unknown OS"),
    };

    let factory = GuiFactory::new(os);

    render_factory(&factory);
}
